### PR TITLE
Unify Process Values

### PR DESCRIPTION
**Pull request by rafabailon**
# Description

| Related Issue |
| --- |
| https://github.com/wazuh/wazuh-qa/issues/5479 |

Create a method in the `DataVisualizer` class that allows to unify the data obtained from the processes and subprocesses so that, when a plot is generated, the data of processes and subprocesses do not appear but all the data unified in a single process.

---

## Testing performed

The tests can be performed locally using the `DataVisualizer` or using the [CLUSTER-Workload_benchmarks_metrics](https://ci.wazuh.info/job/CLUSTER-Workload_benchmarks_metrics/) pipeline with the simplest possible tests. It is only necessary to activate the option to unify data and provide 2 or more CSV of processes (one process and one or more sub processes).

| Validation | Jenkins | Local | OS |
| --- | --- | --- | --- |
| Check plots and painting information | 🟢  | 🟢 | Ubuntu 22.04 |

- Build: https://ci.wazuh.info/view/Tests_DEV/job/CLUSTER-Workload_benchmarks_metrics_5479/2/

### CODE REVIEW

**Comment by rafabailon** [comment]
## Local Test

<details>
   <summary>CSV Files</summary>
   <br/>

- [wazuh-indexer.csv](https://github.com/user-attachments/files/15837661/wazuh-indexer.csv)
- [wazuh-indexer_child_1.csv](https://github.com/user-attachments/files/15837666/wazuh-indexer_child_1.csv)

</details>

<details>
   <summary>Steps</summary>
   <br/>

```console
git clone https://github.com/wazuh/wazuh-qa.git
cd wazuh-qa
git checkout enhancement/5479-unify-process-values
python3 -m pip install -r requirements.txt
cd deps/wazuh_testing/
python3 setup.py install
cd wazuh_testing/scripts/
python3 data_visualizations.py -s wazuh-indexer.csv wazuh-indexer_child_1.csv -u True -t binary -d ~
```

</details>

<details>
   <summary>Plots Generated</summary>
   <br/>

image

[results.zip](https://github.com/user-attachments/files/15837680/results.zip)

</details>

**Comment by rafabailon** [comment]
## Jenkins Test

- Build: https://ci.wazuh.info/view/Tests_DEV/job/CLUSTER-Workload_benchmarks_metrics_5479/1/
- Artifacts: [artifacts.zip](https://github.com/user-attachments/files/15838764/artifacts.zip)

<details>
   <summary>Evidences</summary>
   <br/>

image

image

> Note: No CSVs of subprocesses have been generated due to the simplicity of the test. However, it is possible to see that the command has used the `-u` parameter and the plot has been generated correctly.

</details>

File: deps/wazuh_testing/wazuh_testing/tools/performance/visualization.py [hunk 1]
```diff
@@ -32,7 +32,7 @@ class DataVisualizer:
         base_name (str, optional): base name used to store the images.
     """
     def __init__(self, dataframes, target, compare=False, store_path=gettempdir(), x_ticks_granularity='minutes',
-                 x_ticks_interval=1, base_name=None, columns_path=None):
+                 x_ticks_interval=1, base_name=None, columns_path=None, unify=False):
```
Comment (Rebits) [inline deps/wazuh_testing/wazuh_testing/tools/performance/visualization.py:35]:
```suggestion
                 x_ticks_interval=1, base_name=None, columns_path=None, unify_child_daemon_metrics=False):
```

File: deps/wazuh_testing/wazuh_testing/tools/performance/visualization.py [hunk 2]
```diff
@@ -48,6 +48,9 @@ def __init__(self, dataframes, target, compare=False, store_path=gettempdir(), x
         if target in ['binary', 'analysis', 'remote', 'agent', 'logcollector', 'wazuhdb']:
             self.columns_to_plot = self._load_columns_to_plot(columns_path)
 
+        if unify.lower() in ["true"]:
+            self._unify_dataframes()
```
Comment (Rebits) [inline deps/wazuh_testing/wazuh_testing/tools/performance/visualization.py:52]:
```suggestion
        if unify_child_daemons_metrics:
            self._unify_dataframes()
```

**Review by Rebits — CHANGES_REQUESTED** [review]
The proposed logic does not meet the issue requirements. It should unify the binary metrics of each daemon and its child processes. Currently, it unifies the metrics of all the daemons together.

image

File: deps/wazuh_testing/wazuh_testing/tools/performance/visualization.py (hunk 1 above)
Comment (rafabailon) [reply deps/wazuh_testing/wazuh_testing/tools/performance/visualization.py]: Solved in https://github.com/wazuh/wazuh-qa/pull/5500/commits/4f128638fdbf2976fae270c15f42731b033bddf2

File: deps/wazuh_testing/wazuh_testing/tools/performance/visualization.py (hunk 2 above)
Comment (rafabailon) [reply deps/wazuh_testing/wazuh_testing/tools/performance/visualization.py]: Solved in https://github.com/wazuh/wazuh-qa/pull/5500/commits/4f128638fdbf2976fae270c15f42731b033bddf2

**Comment by rafabailon** [comment]
## Update

The problem was that, when creating the dataframe, a grouping by `Timestamp` was established. I have corrected this to be able to correctly group the data and unify them. I have improved the unification conditional to avoid errors due to a bad use. I have also improved the logic with dataframe specific functions. I have taken the decision to keep and unify the `PID` and `Version` values since we can choose which columns to plot and which to discard. This avoids errors because if we do not discard those columns and unify, it would give error when plotting.

## Local Test

<details>
   <summary>CSV Files</summary>
   <br/>

[csv_files.zip](https://github.com/user-attachments/files/15883406/csv_files.zip)

</details>

<details>
   <summary>Steps</summary>
   <br/>

```console
git clone https://github.com/wazuh/wazuh-qa.git
cd wazuh-qa
git checkout enhancement/5479-unify-process-values
python3 -m pip install -r requirements.txt
cd deps/wazuh_testing/
python3 setup.py install
cd wazuh_testing/scripts/
python3 data_visualizations.py -s wazuh-apid.csv wazuh-apid_child_1.csv wazuh-apid_child_2.csv wazuh-analysisd.csv wazuh_clusterd.csv wazuh_clusterd_child_1.csv -u True -t binary -d ~
```

</details>

<details>
   <summary>Plots Generated</summary>
   <br/>

image

image

[plots.zip](https://github.com/user-attachments/files/15883443/plots.zip)

</details>

**Comment by rafabailon** [comment]
## Jenkins Test

- Build: https://ci.wazuh.info/view/Tests_DEV/job/CLUSTER-Workload_benchmarks_metrics_5479/2/
- Artifacts: 

<details>
   <summary>Evidences</summary>
   <br/>



</details>

File: deps/wazuh_testing/wazuh_testing/tools/performance/visualization.py [hunk 3]
```diff
@@ -48,6 +48,10 @@ def __init__(self, dataframes, target, compare=False, store_path=gettempdir(), x
         if target in ['binary', 'analysis', 'remote', 'agent', 'logcollector', 'wazuhdb']:
             self.columns_to_plot = self._load_columns_to_plot(columns_path)
 
+        if unify_child_daemon_metrics.lower() in ["true"] and target in ['binary']:
+            self.dataframe = self.dataframe.reset_index(drop=False)
+            self._unify_dataframes()
```
Comment (Rebits) [inline deps/wazuh_testing/wazuh_testing/tools/performance/visualization.py:53]:
```suggestion
        if unify_child_daemon_metrics:
            if target=='binary':
                self.dataframe = self.dataframe.reset_index(drop=False)
                self._unify_dataframes()
            else:
                  logging.warning("Enabled unify is only available for binary data. Ignoring")
```

File: deps/wazuh_testing/wazuh_testing/tools/performance/visualization.py [hunk 4]
```diff
@@ -87,6 +91,24 @@ def _load_dataframes(self):
                 new_csv = pd.read_csv(df_path, index_col="Timestamp", parse_dates=True)
                 self.dataframe = pd.concat([self.dataframe, new_csv])
 
+    def _unify_dataframes(self):
+        """Unify dataframe values."""
```
Comment (Rebits) [inline deps/wazuh_testing/wazuh_testing/tools/performance/visualization.py:95]: Improve this docstring. It unify dataframe regarding the Daemon value.

File: deps/wazuh_testing/wazuh_testing/scripts/data_visualizations.py [hunk 5]
```diff
@@ -20,6 +20,8 @@ def get_script_arguments():
                         help=f'Base name for the images. Default {None}.')
     parser.add_argument('-c', '--columns', dest='columns', default=None,
                         help=f'Path to Json with Columns to Plot. Default {None}.')
+    parser.add_argument('-u', '--unify', dest='unify', default=False,
+                        help=f'Unify data of the binary processes with their subprocesses to plot. Default {False}.')
```
Comment (Rebits) [inline deps/wazuh_testing/wazuh_testing/scripts/data_visualizations.py:24]: Currently this require to provide a true/false string value. 

It's more usable using store_true action 

```suggestion
    parser.add_argument('-u', '--unify', dest='unify', default=False, action='store_true',
                        help=f'Unify data of the binary processes with their subprocesses to plot. Default {False}.')
```

**Review by Rebits — CHANGES_REQUESTED** [review]
Some final changes

File: deps/wazuh_testing/wazuh_testing/tools/performance/visualization.py (hunk 4 above)
Comment (rafabailon) [reply deps/wazuh_testing/wazuh_testing/tools/performance/visualization.py]: Solved in https://github.com/wazuh/wazuh-qa/pull/5500/commits/ee85a1dc18d61450298897d04af637de0c4f2051

File: deps/wazuh_testing/wazuh_testing/scripts/data_visualizations.py (hunk 5 above)
Comment (rafabailon) [reply deps/wazuh_testing/wazuh_testing/scripts/data_visualizations.py]: Solved in https://github.com/wazuh/wazuh-qa/pull/5500/commits/ee85a1dc18d61450298897d04af637de0c4f2051

**Review by Rebits — DISMISSED** [review]
LGTM, gj

**Review by juliamagan — APPROVED** [review]

**Pull request merged**